feat(agent): support deleting dev containers

DanielleMaywood opened the pull request:
Relates to https://github.com/coder/coder/issues/19062

Add logic to the agent to allow deleting Dev Containers

---

PR Stack:
- [ ] https://github.com/coder/coder/pull/21247 (Current PR 🚨)
- [ ] https://github.com/coder/coder/pull/21248
- [ ] https://github.com/coder/coder/pull/21249

DanielleMaywood committed:
feat(agent): support deleting dev containers

mafredri commented on agent/agentcontainers/api.go:
```diff
@@ -1224,6 +1229,137 @@ func (api *API) getContainers() (codersdk.WorkspaceAgentListContainersResponse,
+				Detail:  err.Error(),
+			})
+			return
+		}
```
Error paths seem to be lacking state handling for the devcontainer. Should probably update `dc.Error`, perhaps others?

mafredri reviewed (commented):
I did a quick review, didn't really dig into the tests yet.

DanielleMaywood committed:
chore: Stopping -> Deleting

DanielleMaywood committed:
chore: add stopping phase

DanielleMaywood commented on agent/agentcontainers/api.go:
```diff
@@ -1224,6 +1229,137 @@ func (api *API) getContainers() (codersdk.WorkspaceAgentListContainersResponse,
+	delete(api.recreateSuccessTimes, workspaceFolder)
+	delete(api.recreateErrorTimes, workspaceFolder)
+	delete(api.usingWorkspaceFolderName, workspaceFolder)
+	delete(api.injectedSubAgentProcs, workspaceFolder)
```
The purpose of this PR is to explicitly support _removing_ it, not stopping it (as per https://github.com/coder/coder/issues/19062)

DanielleMaywood committed:
chore: appease linter

DanielleMaywood committed:
chore: remove unneeded variable

johnstcn commented on agent/agentcontainers/api.go:
```diff
@@ -1224,6 +1232,164 @@ func (api *API) getContainers() (codersdk.WorkspaceAgentListContainersResponse,
+
+	// NOTE(DanielleMaywood):
+	// We currently do not support canceling the startup of a dev container.
+	if dc.Status == codersdk.WorkspaceAgentDevcontainerStatusStarting {
```
suggestion: Would it make sense to define a method `IsTerminal()` on the `WorkspaceAgentDevcontainerStatus` type that returns true for any *-ing state? My understanding is that we don't want to allow any modifications to *-ing devcontainers.

DanielleMaywood committed:
chore: some comments

DanielleMaywood committed:
refactor: add `IsTransitional` method

We don't allow transitioning a Dev Container when it is currently being
transitioned

mafredri commented on agent/agentcontainers/api.go:
```diff
@@ -1224,6 +1234,158 @@ func (api *API) getContainers() (codersdk.WorkspaceAgentListContainersResponse,
+		api.mu.Unlock()
+
+		httpapi.Write(ctx, w, http.StatusConflict, codersdk.Response{
+			Message: "Unable to delete transitioning Devcontainer",
```
```suggestion
			Message: "Unable to delete transitioning devcontainer",
```

mafredri commented on agent/agentcontainers/api.go:
```diff
@@ -1224,6 +1234,158 @@ func (api *API) getContainers() (codersdk.WorkspaceAgentListContainersResponse,
+			api.mu.Unlock()
+
+			httpapi.Write(ctx, w, http.StatusInternalServerError, codersdk.Response{
+				Message: "An internal error occurred stopping the container",
```
```suggestion
				Message: "An error occurred stopping the container",
```
Nit: Internal suggests fault in api.

mafredri commented on agent/agentcontainers/api.go:
```diff
@@ -1224,6 +1234,158 @@ func (api *API) getContainers() (codersdk.WorkspaceAgentListContainersResponse,
+			api.mu.Unlock()
+
+			httpapi.Write(ctx, w, http.StatusInternalServerError, codersdk.Response{
+				Message: "An internal error occurred removing the container",
```
```suggestion
				Message: "An error occurred removing the container",
```

mafredri commented on agent/agentcontainers/api.go:
```diff
@@ -1240,28 +1402,18 @@ func (api *API) handleDevcontainerRecreate(w http.ResponseWriter, r *http.Reques
 		httpapi.Write(ctx, w, http.StatusConflict, codersdk.Response{
-			Message: "Devcontainer recreation already in progress",
-			Detail:  fmt.Sprintf("Recreation for devcontainer %q is already underway.", dc.Name),
+			Message: "Unable to recreate transitioning Devcontainer",
```
```suggestion
			Message: "Unable to recreate transitioning devcontainer",
```

DanielleMaywood committed:
chore: modify error messages

DanielleMaywood committed:
chore: IsTransitional -> Transitioning

DanielleMaywood committed:
refactor: lock placements, status updates and test helpers
